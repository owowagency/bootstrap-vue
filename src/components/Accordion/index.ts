import Accordion from './Accordion.vue';

export default Accordion;
