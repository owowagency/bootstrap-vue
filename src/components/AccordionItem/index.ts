import AccordionItem from './AccordionItem.vue';

export default AccordionItem;
