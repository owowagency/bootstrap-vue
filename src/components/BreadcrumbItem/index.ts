import BreadcrumbItem from './BreadcrumbItem.vue';

export default BreadcrumbItem;
