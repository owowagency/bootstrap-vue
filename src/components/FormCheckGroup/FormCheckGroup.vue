<template>
    <FormCheck
        v-for="(item, index) of items"
        :key="item.value"
        :disabled="disabled"
        :label="item.label || item.value"
        :model-value="getValue(index)"
        :name="name"
        :type="type"
        :switch="isSwitch"
        :value="item.value"
        @update:model-value="setValue($event, index)"
    />
</template>

<script lang="ts">
import FormCheck, {Type, typeProp} from '@/components/FormCheck';
import {PropType, computed} from 'vue';
import {disabledProps} from '@/composables/useDisabled';
import {idProps} from '@/composables/useId';
</script>

<script lang="ts" setup>
interface Item {
    label?: string;
    value: unknown;
}

const props = defineProps({
    ...disabledProps,
    /**
     * Display the checks as a switch by adding the `form-switch` class
     */
    switch: {
        type: Boolean,
        default: false,
    },
    /**
     * The items to display in the group
     */
    items: {
        type: Array as PropType<Item[]>,
        required: true,
    },
    /**
     * The type of the input
     * @values 'checkbox'|'radio'
     */
    type: typeProp,
    // Put `modelValue` after `type` so it will be available in `default`.
    /**
     * The model value
     */
    modelValue: {
        type: [Array, String, Number, undefined],
        default: (props: {type: Type}) => props.type === 'checkbox' ? [] : undefined,
    },
});

const name = props.type === 'checkbox' ? undefined : idProps.id.default();

const emit = defineEmits([
    /**
     * Fired when the model value is updated
     * @param {any} value The new value
     */
    'update:modelValue',
]);

// Required to prevent the use of reserved word `switch` in template.
const isSwitch = computed(() => props.switch);

const value = computed<unknown[]|string|number|undefined>({
    get: () => props.modelValue,
    set: (v) => emit('update:modelValue', v),
});

const getValue = (i) => props.type === 'checkbox' ? !!value.value[i] : value.value;

const setValue = (v, i) => {
    if (props.type === 'checkbox') {
        const clone = [...value.value as unknown[]];

        clone[i] = v;

        value.value = clone;
    } else {
        value.value = v;
    }
};
</script>

<docs>
```vue
<FormCheckGroup
    v-model="checks"
    :items="[Item]"
    type="checkbox"
/>
```

## Imported Props

| Prop name | Description            | Type    | Values | Default |
| --------- | ---------------------- | ------- | ------ | ------- |
| disabled  | Disables the component | boolean | -      | false   |
</docs>
