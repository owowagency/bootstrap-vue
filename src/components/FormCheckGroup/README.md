# FormCheckGroup

```vue
<FormCheckGroup v-model="checks" :items="[Item]" type="checkbox" />
```

## Imported Props

| Prop name | Description            | Type    | Values | Default |
| --------- | ---------------------- | ------- | ------ | ------- |
| disabled  | Disables the component | boolean | -      | false   |

## Props

| Prop name  | Description                                                      | Type                             | Values                | Default                                                                |
| ---------- | ---------------------------------------------------------------- | -------------------------------- | --------------------- | ---------------------------------------------------------------------- |
| switch     | Display the checks as a switch by adding the `form-switch` class | boolean                          | -                     | false                                                                  |
| items      | The items to display in the group                                | Item[]                           | -                     |                                                                        |
| type       | The type of the input                                            | typeProp                         | `'checkbox'\|'radio'` |                                                                        |
| modelValue | The model value                                                  | array\|string\|number\|undefined | -                     | (props: {type: Type}) =&gt; props.type === 'checkbox' ? [] : undefined |

## Events

| Event name        | Properties                      | Description                           |
| ----------------- | ------------------------------- | ------------------------------------- |
| update:modelValue | **value** `any` - The new value | Fired when the model value is updated |
