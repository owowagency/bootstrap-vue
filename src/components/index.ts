export {default as Accordion} from './Accordion';
export {default as AccordionItem} from './AccordionItem';
export {default as Alert} from './Alert';
export {default as Badge} from './Badge';
export {default as Breadcrumb} from './Breadcrumb';
export {default as BreadcrumbItem} from './BreadcrumbItem';
export {default as Button} from './Button';
export {default as ButtonClose} from './ButtonClose';
export {default as ButtonGroup} from './ButtonGroup';
export {default as ButtonState} from './ButtonState';
export {default as Card} from './Card';
export {default as Carousel} from './Carousel';
export {default as Col} from './Col';
export {default as Collapse} from './Collapse';
export {default as Confirm} from './Confirm';
export {default as Container} from './Container';
export {default as Dropdown} from './Dropdown';
export {default as DropdownMenu} from './DropdownMenu';
export {default as DropdownMenuItem} from './DropdownMenuItem';
export {default as FormCheck} from './FormCheck';
export {default as FormCheckGroup} from './FormCheckGroup';
export {default as FormControl} from './FormControl';
export {default as FormDropdown} from './FormDropdown';
export {default as FormDropdownSearch} from './FormDropdownSearch';
export {default as FormGroup} from './FormGroup';
export {default as FormImage} from './FormImage';
export {default as FormPills} from './FormPills';
export {default as FormSelect} from './FormSelect';
export {default as InputGroup} from './InputGroup';
export {default as InputPassword} from './InputPassword';
export {default as ListGroup} from './ListGroup';
export {default as ListGroupItem} from './ListGroupItem';
export {default as Modal} from './Modal';
export {default as Nav} from './Nav';
export {default as NavItem} from './NavItem';
export {default as NavLink} from './NavLink';
export {default as Offcanvas} from './Offcanvas';
export {default as Row} from './Row';
export {default as Spinner} from './Spinner';
export {default as Table} from './Table';
