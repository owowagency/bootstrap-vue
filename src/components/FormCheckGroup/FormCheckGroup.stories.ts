import FormCheckGroup from '.';
import {ref} from 'vue';
import {types} from '@/components/FormCheck';

export default {
    title: 'Forms/Form check group',
    component: FormCheckGroup,
    argTypes: {
        disabled: {
            control: {type: 'boolean'},
        },
        switch: {
            control: {type: 'boolean'},
        },
        type: {
            control: {type: 'select'},
            options: types,
        },
    },
};

const items = [
    {label: 'Item one', value: 'one'},
    {label: 'Item two', value: 'two'},
];

const template = (args) => ({
    components: {FormCheckGroup},
    setup: () =>({args, value: ref([])}),
    template: `
        <FormCheckGroup v-model="value" v-bind="args" />
    `,
});

export const Checkbox = template.bind({});

Checkbox.args = {
    items,
    type: 'checkbox',
};

export const Radio = template.bind({});

Radio.args = {
    items,
    type: 'radio',
};

export const Disabled = template.bind({});

Disabled.args = {
    disabled: true,
    items,
};

export const Switch = template.bind({});

Switch.args = {
    switch: true,
    items,
};
