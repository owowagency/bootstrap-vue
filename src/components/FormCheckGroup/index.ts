import FormCheckGroup from './FormCheckGroup.vue';

export default FormCheckGroup;
