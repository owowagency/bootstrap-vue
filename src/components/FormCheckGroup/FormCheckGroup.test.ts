import {describe, expect, it, vi} from 'vitest';
import {FormCheck} from '@/components';
import FormCheckGroup from '.';
import {shallowMount} from '@vue/test-utils';

// Mock uuid module because it is used as `id` in `FormCheck` and we are unable
// to reach that property directly.
vi.mock('uuid', () => ({
    v4: () => 'wordpres-sist-hebe-stwe-bsolution!!!',
}));

const items = [
    {label: 'Item one', value: 'one'},
    {value: 'two'},
];

describe('template', () => {
    componentRenderTest(FormCheckGroup, {props: {items}});

    componentRenderTest(
        FormCheckGroup,
        {props: {items, type: 'radio'}},
        true,
        'renders radio',
    );

    it('emits on update:modelValue of child', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'checkbox'}},
        );

        wrapper.findComponent(FormCheck)
            .vm
            .$emit('update:modelValue', 'one');

        expect(wrapper.emitted('update:modelValue')[0]).toEqual([['one']]);
    });
});

describe('modelValue', () => {
    it('is an array by default when is type checkbox', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'checkbox'}},
        );

        expect(wrapper.vm.modelValue).toEqual([]);
    });

    it('is undefined by default when is type radio', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'radio'}},
        );

        expect(wrapper.vm.modelValue).toBeUndefined();
    });
});

describe('name', () => {
    it('is undefined when type is checkbox', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'checkbox'}},
        );

        expect(wrapper.vm.name).toBeUndefined();
    });

    it('is uuid when type is radio', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'radio'}},
        );

        expect(wrapper.vm.name).toBe('owow-wordpres-sist-hebe-stwe-bsolution!!!');
    });
});

describe('isSwitch', () => {
    it('is equal to switch', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, switch: true}},
        );

        expect(wrapper.vm.switch).toBe(true);
    });
});

describe('value', () => {
    it('is equal to modelValue', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, modelValue: ['two']}},
        );

        expect(wrapper.vm.value).toEqual(['two']);
    });

    it('emits update:modelValue when set', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items}},
        );

        wrapper.vm.value = ['two'];

        expect(wrapper.emitted('update:modelValue')[0]).toEqual([['two']]);
    });
});

describe('getValue', () => {
    describe('when checkbox', () => {
        it('returns true when in value array', () => {
            const wrapper = shallowMount(
                FormCheckGroup,
                {props: {items, modelValue: ['two'], type: 'checkbox'}},
            );

            expect(wrapper.vm.getValue(0)).toBe(true);
        });

        it('returns false when not in value array', () => {
            const wrapper = shallowMount(
                FormCheckGroup,
                {props: {items, modelValue: ['two'], type: 'checkbox'}},
            );

            expect(wrapper.vm.getValue(1)).toBe(false);
        });
    });

    describe('when radio', () => {
        it('returns value', () => {
            const wrapper = shallowMount(
                FormCheckGroup,
                {props: {items, modelValue: 'two', type: 'radio'}},
            );

            expect(wrapper.vm.getValue(0)).toEqual('two');
        });
    });
});

describe('setValue', () => {
    it('sets value in array when checkbox', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'checkbox'}},
        );

        wrapper.vm.setValue('two', 1);

        // It is not possible to assert `value` since it is a computed getter
        // and setter. Eventually it will emit `update:modelValue` when being
        // set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual([[undefined, 'two']]);
    });

    it('sets value in array when checkbox', () => {
        const wrapper = shallowMount(
            FormCheckGroup,
            {props: {items, type: 'radio'}},
        );

        wrapper.vm.setValue('one', 0);

        // It is not possible to assert `value` since it is a computed getter
        // and setter. Eventually it will emit `update:modelValue` when being
        // set.
        expect(wrapper.emitted('update:modelValue')[0]).toEqual(['one']);
    });
});
