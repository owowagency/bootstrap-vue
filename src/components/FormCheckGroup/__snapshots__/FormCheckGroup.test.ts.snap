// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`template > renders default 1`] = `
"<form-check-stub disabled=\\"false\\" id=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" inline=\\"false\\" label=\\"Item one\\" modelvalue=\\"false\\" switch=\\"false\\" type=\\"checkbox\\" value=\\"one\\"></form-check-stub>
<form-check-stub disabled=\\"false\\" id=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" inline=\\"false\\" label=\\"two\\" modelvalue=\\"false\\" switch=\\"false\\" type=\\"checkbox\\" value=\\"two\\"></form-check-stub>"
`;

exports[`template > renders radio 1`] = `
"<form-check-stub disabled=\\"false\\" id=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" inline=\\"false\\" label=\\"Item one\\" modelvalue=\\"false\\" name=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" switch=\\"false\\" type=\\"radio\\" value=\\"one\\"></form-check-stub>
<form-check-stub disabled=\\"false\\" id=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" inline=\\"false\\" label=\\"two\\" modelvalue=\\"false\\" name=\\"owow-wordpres-sist-hebe-stwe-bsolution!!!\\" switch=\\"false\\" type=\\"radio\\" value=\\"two\\"></form-check-stub>"
`;
